### PR TITLE
Fix early evaluation logic for address node of BCDCHK

### PR DESCRIPTION
**Pull request by VermaSh**
Reference count check was added to avoid commoning of dataAddr pointer (first child of address node) by preventing early evaluation. However, this might create functional issues if a later instruction expects the child, and/or siblings of dataAddr pointer to be evaluated into a register. Checking the reference count individually prevents commoning of dataAddr pointer while maintaining functional correctness.

### CODE REVIEW

**Comment by VermaSh** [comment]
@r30shah This is ready for review. Don't see any failures during my local testing with Medicare. Have launched a [personal build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/27203/) with off-heap enabled, but don't expect there to be any new failures due to this.

**Comment by VermaSh** [comment]
Don't see any related failures. Let me just verify trees with 0 offset.

**Comment by VermaSh** [comment]
The change is good for 0 offset as well. @r30shah can i please get a review?

**Comment by VermaSh** [comment]
@zl-wang Can I get a review from off-heap perspective?

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp [hunk 1]
```diff
@@ -2070,10 +2070,47 @@ J9::Z::TreeEvaluator::BCDCHKEvaluatorImpl(TR::Node * node,
...
+            cg->evaluate(secondChild->getFirstChild()->getFirstChild());
+            }
+         }
+      else if (secondChild->getFirstChild()->getReferenceCount() > 1)
+         {
+         cg->evaluate(secondChild->getFirstChild());
+         }
+
+      // Check reference count of offset node to determine if it qualifies for early evaluation
+      if (secondChild->getSecondChild()->getReferenceCount() > 1)
```
Comment (r30shah) [inline runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2110]: What if the `offset` node has only one reference count ?

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp [hunk 2]
```diff
@@ -2070,10 +2070,36 @@ J9::Z::TreeEvaluator::BCDCHKEvaluatorImpl(TR::Node * node,
       callNode->setAndIncChild(i, childRootNode->getChild(i + callChildStartIndex));
 
    // Evaluate secondChild's children, if the secondChild is an address node into a byte[]
-   if(isResultPD && secondChild->getNumChildren() == 2 && secondChild->getReferenceCount() > 1)
-      {
-      cg->evaluate(secondChild->getFirstChild());
-      cg->evaluate(secondChild->getSecondChild());
+   if(isResultPD && secondChild->getNumChildren() == 2)
```
Comment (r30shah) [inline runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2073]: Would secondChild not always be `aladd` node ? I personally found it bit confusing just code-readability point of view, and would have chose different name. Also would have checked the reference count of the children before evaluating it (If they are no longer referenced after this node, would still be OK to delay the evaluation in OOL I think. 

Given the change is blocking a functional issue, I am OK with it.

**Review by r30shah — APPROVED** [review]

**Comment by r30shah** [comment]
Jenkins test sanity zlinux jdk17,jdk21

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp [hunk 3]
```diff
@@ -2070,10 +2070,36 @@ J9::Z::TreeEvaluator::BCDCHKEvaluatorImpl(TR::Node * node,
...
+      {
```
Comment (zl-wang) [inline runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2074]: could you expect there is always an `offset` child?  how about `array[0] = something`? in off-heap mode, there is no offset at all.  i.e. for that kind of tree, code below will crash with NULL child pointer.

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp (hunk 3 above)
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2074]: Yup, that's what I was worried about as well [here](https://github.com/eclipse-openj9/openj9/pull/21454#issuecomment-2748606519), however, we add a const 0 tree as offset. But I suppose the const 0 tree could be optimized out causing us to not enter the if block all together. So we might crash due to arrayObject not being evaluated. I didn't see this happen during my verification runs, maybe I just need to run more iterations to hit this edge case. I'll try to get a unit test for this.

**Comment by VermaSh** [comment]
Launched new off-heap personal [build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/27264/) as my last personal build had the reference count changes.

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp (hunk 3 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2074]: Just to make sure I understand clearly - the `offset` node that we are talking about here is the `secondChild` of the `secondChild` (This is the reason I originally commented the confusing variable name used here).  which is referenced and evaluated at line `2102`?

Comment (VermaSh) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2074]: Got a trace log for the corner case with older build. tree simplification replaces `aladd` (`n902n`) containing const 0 tree with `aloadi` (`n901n`) hoisted outside of `BCDCHK`. So we shouldn't see any crashes with the change in this PR, but, I'll re-run the original failing microbenchmark for verification.

**Before treeSimplification**
```
n395n     BBStart <block_25> (freq 7001)                                                      [     0x3fcc8586b30] bci=[2,131,805] rc=0 vc=156 vn=- li=-1 udi=- nc=0
n909n     BCDCHK [#507] ()                                                                    [     0x3fcc86e0be0] bci=[2,0,790] rc=0 vc=156 vn=- li=-1 udi=- nc=12 flg=0x20
n907n       pdshlOverflow <prec=1 (len=1) adj=0 round=0> sign=?                               [     0x3fcc86e0b40] bci=[-1,0,8] rc=2 vc=156 vn=- li=- udi=- nc=2
n899n         pdrem <prec=2 (len=2) adj=0 round=0> sign=?                                     [     0x3fcc86e08c0] bci=[-1,0,8] rc=1 vc=156 vn=- li=- udi=- nc=2
n894n           pdloadi  <array-shadow>[#260  Shadow] [flags 0x80000609 0x0 ] <prec=4 (len=3) adj=0> sign=hasState   [     0x3fcc86e0730] bci=[-1,0,8] rc=1 vc=156 vn=- li=- udi=- nc=1
n893n             aladd (internalPtr )                                                        [     0x3fcc86e06e0] bci=[-1,0,8] rc=1 vc=156 vn=- li=- udi=- nc=2 flg=0x8000
n892n               aloadi  <contiguousArrayDataAddrField>[#355  final Shadow +8] [flags 0x20604 0x0 ] (internalPtr )  [     0x3fcc86e0690] bci=[2,134,805] rc=1 vc=156 vn=- li=- udi=- nc=1 flg=0x8000
n591n                 aload  <auto slot 1>[#400  Auto] [flags 0x7 0x0 ] (X!=0 )               [     0x3fcc858a870] bci=[2,134,805] rc=2 vc=156 vn=- li=- udi=52 nc=0 flg=0x4
n891n               lconst 0 (highWordZero X==0 X>=0 X<=0 )                                   [     0x3fcc86e0640] bci=[-1,61,28] rc=1 vc=156 vn=- li=- udi=- nc=0 flg=0x4302
n898n           pdloadi  <array-shadow>[#260  Shadow] [flags 0x80000609 0x0 ] <prec=2 (len=2) adj=0> sign=hasState   [     0x3fcc86e0870] bci=[-1,0,8] rc=1 vc=156 vn=- li=- udi=- nc=1
n897n             aladd (internalPtr )                                                        [     0x3fcc86e0820] bci=[-1,0,8] rc=1 vc=156 vn=- li=- udi=- nc=2 flg=0x8000
n896n               aloadi  <contiguousArrayDataAddrField>[#355  final Shadow +8] [flags 0x20604 0x0 ] (internalPtr )  [     0x3fcc86e07d0] bci=[2,139,805] rc=1 vc=156 vn=- li=- udi=- nc=1 flg=0x8000
n594n                 aload  <auto slot 2>[#401  Auto] [flags 0x7 0x0 ] (X!=0 )               [     0x3fcc858a960] bci=[2,139,805] rc=2 vc=156 vn=- li=- udi=54 nc=0 flg=0x4
n895n               lconst 0 (highWordZero X==0 X>=0 X<=0 )                                   [     0x3fcc86e0780] bci=[-1,65,28] rc=1 vc=156 vn=- li=- udi=- nc=0 flg=0x4302
n906n         iconst 0 (X==0 X>=0 X<=0 )                                                      [     0x3fcc86e0af0] bci=[2,147,805] rc=1 vc=156 vn=- li=- udi=- nc=0 flg=0x302
n902n       aladd (internalPtr )                                                              [     0x3fcc86e09b0] bci=[-1,0,8] rc=1 vc=156 vn=- li=- udi=- nc=2 flg=0x8000
n901n         aloadi  <contiguousArrayDataAddrField>[#355  final Shadow +8] [flags 0x20604 0x0 ] (internalPtr )  [     0x3fcc86e0960] bci=[2,131,805] rc=1 vc=156 vn=- li=- udi=- nc=1 flg=0x8000
n588n           aload  <auto slot 3>[#402  Auto] [flags 0x7 0x0 ] (X!=0 )                     [     0x3fcc858a780] bci=[2,131,805] rc=3 vc=156 vn=- li=- udi=50 nc=0 flg=0x4
n900n         lconst 0 (highWordZero X==0 X>=0 X<=0 )                                         [     0x3fcc86e0910] bci=[-1,57,28] rc=1 vc=156 vn=- li=- udi=- nc=0 flg=0x4302
n588n       ==>aload
```
**After treeSimplification**
```
<optimization id=47 name=treeSimplification method=TestArithmeticOperations.testRemainderPrecisionTooSmallOverflowSetAsTrueEvenToOdd()V>
Performing 47: treeSimplification
...
[   277] O^O TREE SIMPLIFICATION: Replace node [000003FCC86E09B0] aladd by [000003FCC86E0960] aloadi
         O^O TREE SIMPLIFICATION: anchor child aloadi [000003FCC86E0960] at depth 0 before BCDCHK [000003FCC86E0BE0]
...
n387n     BBStart <block_31> (freq 10000)                                                     [     0x3fcc85868b0] bci=[2,131,805] rc=0 vc=195 vn=- li=- udi=- nc=0
n936n     treetop                                                                             [     0x3fcc86e1450] bci=[2,131,805] rc=0 vc=0 vn=- li=- udi=- nc=1
n901n       aloadi  <contiguousArrayDataAddrField>[#355  final Shadow +8] [flags 0x20604 0x0 ] (internalPtr )  [     0x3fcc86e0960] bci=[2,131,805] rc=2 vc=195 vn=- li=- udi=- nc=1 flg=0x8000
n588n         aload  <auto slot 3>[#402  Auto] [flags 0x7 0x0 ] (X!=0 X>=0 )                  [     0x3fcc858a780] bci=[2,131,805] rc=3 vc=195 vn=- li=- udi=50 nc=0 flg=0x104
n937n     treetop                                                                             [     0x3fcc86e14a0] bci=[2,139,805] rc=0 vc=0 vn=- li=- udi=- nc=1
n896n       aloadi  <contiguousArrayDataAddrField>[#355  final Shadow +8] [flags 0x20604 0x0 ] (internalPtr )  [     0x3fcc86e07d0] bci=[2,139,805] rc=2 vc=195 vn=- li=- udi=- nc=1 flg=0x8000
n594n         aload  <auto slot 2>[#401  Auto] [flags 0x7 0x0 ] (X!=0 X>=0 )                  [     0x3fcc858a960] bci=[2,139,805] rc=2 vc=195 vn=- li=- udi=54 nc=0 flg=0x104
n938n     treetop                                                                             [     0x3fcc86e14f0] bci=[2,134,805] rc=0 vc=0 vn=- li=- udi=- nc=1
n892n       aloadi  <contiguousArrayDataAddrField>[#355  final Shadow +8] [flags 0x20604 0x0 ] (internalPtr )  [     0x3fcc86e0690] bci=[2,134,805] rc=2 vc=195 vn=- li=- udi=- nc=1 flg=0x8000
n591n         aload  <auto slot 1>[#400  Auto] [flags 0x7 0x0 ] (X!=0 X>=0 )                  [     0x3fcc858a870] bci=[2,134,805] rc=2 vc=195 vn=- li=- udi=52 nc=0 flg=0x104
n909n     BCDCHK [#507] ()                                                                    [     0x3fcc86e0be0] bci=[2,0,790] rc=0 vc=195 vn=- li=- udi=- nc=12 flg=0x20
n907n       pdshlOverflow <prec=1 (len=1) adj=0 round=0> sign=?                               [     0x3fcc86e0b40] bci=[-1,0,8] rc=2 vc=195 vn=- li=- udi=- nc=2
n899n         pdrem <prec=2 (len=2) adj=0 round=0> sign=?                                     [     0x3fcc86e08c0] bci=[-1,0,8] rc=1 vc=195 vn=- li=- udi=- nc=2
n894n           pdloadi  <array-shadow>[#260  Shadow] [flags 0x80000609 0x0 ] <prec=4 (len=3) adj=0> sign=hasState   [     0x3fcc86e0730] bci=[-1,0,8] rc=1 vc=195 vn=- li=- udi=- nc=1
n892n             ==>aloadi
n898n           pdloadi  <array-shadow>[#260  Shadow] [flags 0x80000609 0x0 ] <prec=2 (len=2) adj=0> sign=hasState   [     0x3fcc86e0870] bci=[-1,0,8] rc=1 vc=195 vn=- li=- udi=- nc=1
n896n             ==>aloadi
n906n         iconst 0 (X==0 X>=0 X<=0 )                                                      [     0x3fcc86e0af0] bci=[2,147,805] rc=1 vc=195 vn=- li=- udi=- nc=0 flg=0x302
n901n       ==>aloadi
```

Comment (VermaSh) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2074]: That being said, I don't think there is any harm in doing early evaluation of arrayObject node if off-heap is enabled.

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp (hunk 2 above)
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2073]: Sorry, completely, missed this comment. With my latest verification run, the `aladd` can be replaced by a `aloadi` if the offset is 0. This would only happen in off-heap mode.

**Comment by VermaSh** [comment]
I have updated the PR with changes to catch the edge case and have updated the variable name to make it . Can I please get another review @r30shah , @zl-wang

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp [hunk 4]
```diff
@@ -2069,11 +2069,46 @@ J9::Z::TreeEvaluator::BCDCHKEvaluatorImpl(TR::Node * node,
...
+         {
+         TR_ASSERT_FATAL_WITH_NODE(node,
+            addressNode->getFirstChild()->getReferenceCount() <= 1,
+            "DataAddr pointer isn't being commoned so it shouldn't have reference count greater than 1.\n");
+
+         arrayObjectNode = addressNode->getFirstChild()->getFirstChild();
+         }
+
+      cg->evaluate(arrayObjectNode);
+
```
Comment (zl-wang) [inline runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2108]: this is going to prevent MemoryReference constructor from exploiting the best instruction to use (i.e. [baseAddr, constOffset]  and  [baseAddr, indexReg, constOffset] addressing modes), since the offset subtree is now evaluated into a register. look at x or p side changes to get some ideas.

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp [hunk 5]
```diff
@@ -2069,11 +2069,46 @@ J9::Z::TreeEvaluator::BCDCHKEvaluatorImpl(TR::Node * node,
...
+       *     aladd
+       *      arrayObject
+       *      offset
+       *
+       * contiguousArrayDataAddrFieldSymbol isn't being commoned right now so it shouldn't
+       * have reference count > 1; arrayObject and offset nodes are the only nodes we need to
+       * evaluate early.
+       */
+      TR::Node *arrayObjectNode = addressNode->getFirstChild();
+      if (addressNode->getFirstChild()->isDataAddrPointer())
```
Comment (zl-wang) [inline runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2098]: this narrow checking will run into bugs i imagined (applicable to the outer checking too) ... i had similar comments for @rmnattas as well for his x86 change. for a compilation which may handle dataAddr load as internal pointer to begin with, i can envision the dataAddr load being turned into aRegLoad which doesn't have isDataAddrPointer property anymore ... you will run into bugs. currently at ilGen time, i know dataAddr load isn't treated as a full fledged internal pointer yet, but anything prevents hot/scorching compilations from commoning them up and making them into aRegStore and aRegLoad?

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp (hunk 4 above)
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2108]: It gets the constant instead of using the offset register. In this trace log snippet from zlinux, `0x3fcb6343390` uses const 2 even though the offset tree (`n903n`) was already evaluated into a register at `0x3fcb6340140`. So, I think we should we good here.

```
------------------------------
 n911n    (  0)  BCDCHK [#465] (profilingCode )                                                       [     0x3fcb62e0c80] bci=[1,39,281] rc=0 vc=415 vn=- li=22 udi=- nc=7 flg=0xa0
 n910n    (  1)    pdshlOverflow <prec=2 (len=2) adj=0 round=0> sign=?  (in VRF_0261) (profilingCode )  [     0x3fcb62e0c30] bci=[1,39,281] rc=1 vc=415 vn=- li=-1 udi=- nc=2 flg=0x80
 n902n    (  0)      i2pd <prec=2 (len=2) srcprec=2> sign=?  (in VRF_0261) (profilingCode X>=0 )      [     0x3fcb62e09b0] bci=[1,39,281] rc=0 vc=415 vn=- li=- udi=- nc=1 flg=0x180
 n32n     (  0)        iconst 11 (in GPR_0256) (profilingCode X!=0 X>=0 )                             [     0x3fcb6105580] bci=[-1,23,16] rc=0 vc=415 vn=- li=22 udi=- nc=0 flg=0x184
 n28n     (  5)      ==>iconst 0 (in GPR_0179) (profilingCode X==0 X>=0 X<=0 )
 n905n    (  0)    aladd (in GPR_0303) (profilingCode internalPtr )                                   [     0x3fcb62e0aa0] bci=[-1,0,8] rc=0 vc=415 vn=- li=22 udi=- nc=2 flg=0x8080
 n904n    (  0)      aloadi  <contiguousArrayDataAddrField>[#355  final Shadow +8] [flags 0x20604 0x0 ] (in &GPR_0258) (profilingCode internalPtr )  [     0x3fcb62e0a50] bci=[1,40,281] r
 n20n     (  4)        ==>newarray (in &GPR_0082) (highWordZero Unsigned profilingCode X!=0 X>=0 allocationCanBeRemoved )
 n903n    (  1)      lconst 2 (in GPR_0260) (highWordZero profilingCode X!=0 X>=0 )                   [     0x3fcb62e0a00] bci=[-1,49,24] rc=1 vc=415 vn=- li=22 udi=- nc=0 flg=0x4184
 n32n     (  0)    ==>iconst 11 (in GPR_0256) (profilingCode X!=0 X>=0 )
 n20n     (  4)    ==>newarray (in &GPR_0082) (highWordZero Unsigned profilingCode X!=0 X>=0 allocationCanBeRemoved )
 n36n     (  1)    iconst 2 (in GPR_0257) (profilingCode X!=0 X>=0 )                                  [     0x3fcb61056c0] bci=[-1,30,19] rc=1 vc=415 vn=- li=22 udi=- nc=0 flg=0x184
 n36n     (  1)    ==>iconst 2 (in GPR_0257) (profilingCode X!=0 X>=0 )
 n28n     (  5)    ==>iconst 0 (in GPR_0179) (profilingCode X==0 X>=0 X<=0 )
------------------------------

 [     0x3fcb633f970]                          LHI     GPR_0256,0xb
 [     0x3fcb633fac0]                          LHI     GPR_0257,0x2
 [     0x3fcb6340000]                          LTG     &GPR_0258, Shadow[<contiguousArrayDataAddrField>] 8(&GPR_0082)
 [     0x3fcb6340140]                          LGHI    GPR_0260,0x2
 [     0x3fcb6340290]                          VCVD    VRF_0261,GPR_0256,0x2,1
 [     0x3fcb6340440]                          BRC     BRO(0x0), Snippet Label L0098
 [     0x3fcb6340520]                          BRC     BRO(0x0), Outlined Label L0096
 [     0x3fcb6343cc0]                          assocreg
 [     0x3fcb6343700]                          Label L0097:
 POST:
 {AssignAny:VRF_0261:R} {AssignAny:GPR_0256:R}

------------------------------

 [     0x3fcb6340650]                          Outlined Label L0096:     ; Start of BCDCHK OOL sequence
 [     0x3fcb63413c0]                          LGR     GPR_0262,GPR_0256         ; LR=Reg_param
 [     0x3fcb6341530]                          LGR     &GPR_0263,&GPR_0082       ; LR=Reg_param
 [     0x3fcb63416a0]                          LGR     GPR_0264,GPR_0257         ; LR=Reg_param
 [     0x3fcb6341860]                          MVHI    #542 8(GPR5),0x0002
 [     0x3fcb6341a00]                          MVHI    #543 0(GPR5),0x0000
 [     0x3fcb6343170]                          assocreg
  PRE:
 {GPR1:GPR_0262:R} {GPR2:&GPR_0263:R} {GPR3:GPR_0264:R}
 [     0x3fcb6342b80]                          BRASL   GPR_0270,0x0000000000000000
 POST:
 {GPR0:D_GPR_0265:D}* {GPR1:D_GPR_0266:D}* {GPR2:D_GPR_0267:D}* {GPR3:D_GPR_0268:D}* {GPR4:D_GPR_0269:D}* {GPR14:GPR_0270:D} {FPR0:D_FPR_0271:D}* {FPR1:D_FPR_0272:D}*
 {FPR2:D_FPR_0273:D}* {FPR3:D_FPR_0274:D}* {FPR4:D_FPR_0275:D}* {FPR5:D_FPR_0276:D}* {FPR6:D_FPR_0277:D}* {FPR7:D_FPR_0278:D}* {FPR8:D_FPR_0279:D}* {FPR9:D_FPR_0280:D}*
 {FPR10:D_FPR_0281:D}* {FPR11:D_FPR_0282:D}* {FPR12:D_FPR_0283:D}* {FPR13:D_FPR_0284:D}* {FPR14:D_FPR_0285:D}* {FPR15:D_FPR_0286:D}* {VRF16:D_VRF_0287:D}* {VRF17:D_VRF_0288:D}*
 {VRF18:D_VRF_0289:D}* {VRF19:D_VRF_0290:D}* {VRF20:D_VRF_0291:D}* {VRF21:D_VRF_0292:D}* {VRF22:D_VRF_0293:D}* {VRF23:D_VRF_0294:D}* {VRF24:D_VRF_0295:D}* {VRF25:D_VRF_0296:D}*
 {VRF26:D_VRF_0297:D}* {VRF27:D_VRF_0298:D}* {VRF28:D_VRF_0299:D}* {VRF29:D_VRF_0300:D}* {VRF30:D_VRF_0301:D}* {VRF31:D_VRF_0302:D}*
 [     0x3fcb6343390]                          LA      GPR_0303,#544 2(&GPR_0258)
```

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp (hunk 5 above)
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2098]: >but anything prevents hot/scorching compilations from commoning them up and making them into aRegStore and aRegLoad?

dataAddr load would have to be marked as auto/param for it to be transformed to regStore/regLoad. Don't think that is the case at the moment.

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp (hunk 4 above)
Comment (zl-wang) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2108]: this probably is not a good representative case, since that `aladd` seemed required to be evaluated into a register, instead of being able to make into a typical architectural addressing format.

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp (hunk 5 above)
Comment (zl-wang) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2098]: optimizer is supposed to make localization of them, as if GRA candidates were local temps before being transformed into aRegLoad/Store.  so, the question is:  is dataAddr load currently explicitly excluded from/disabled for localization?

Comment (VermaSh) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2098]: I can check, what would be the best place to verify this?
But if dataAddr can be transformed then don't similar checks on [p](https://github.com/eclipse-openj9/openj9/blob/22d04e7f62749bc2424159100f71665f67870e10/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L8315),[aarch64](https://github.com/eclipse-openj9/openj9/blob/22d04e7f62749bc2424159100f71665f67870e10/runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp#L5106) and [x](https://github.com/eclipse-openj9/openj9/blob/22d04e7f62749bc2424159100f71665f67870e10/runtime/compiler/x/codegen/J9TreeEvaluator.cpp#L3034-L3036) which use `->isDataAddrPointer()` also run the risk of failing?

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp (hunk 4 above)
Comment (VermaSh) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2108]: To help me clarify, could you please describe what the IL would look like for the scenario you're referring to?

Comment (zl-wang) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2108]: `aladd` is typically under a load or store node, providing the datum addressing info. your example is atypical in the sense that `aladd` is not under load/store node and appears to be an independent argument to BCDCHK.

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp (hunk 5 above)
Comment (zl-wang) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2098]: from the fact we knew: stackWalker needs to handle internal pointer(s) and there were a few GC bugs re displacement calculation. i am pretty sure dataAddr load or aladd indeed could be localized.

File: runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp (hunk 4 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2108]: @VermaSh  What do we generate for pdloadi - I would have assume what Julian is talking about (Usual reference of the dataAddPtr under load / store - would still be the case for pdloadi/pdstorei right ?

Comment (VermaSh) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2108]: Below is what we do for `pdloadi` and `pdstorei`.
On a related note, treeSimplification seems to be hoisting dataAddr (`n888n`) out of BCDCHK. This wasn't consistent before but seeing it more now, trying to understand what's causing this.
```
------------------------------
 n926n    (  0)  treetop (profilingCode )                                                             [     0x3fcb62e1130] bci=[2,134,805] rc=0 vc=415 vn=- li=31 udi=- nc=1 flg=0x80
 n884n    (  1)    aloadi  <contiguousArrayDataAddrField>[#355  final Shadow +8] [flags 0x20604 0x0 ] (in &GPR_0324) (profilingCode internalPtr )  [     0x3fcb62e0410] bci=[2,134,805] rc=1 vc=415 vn=- li=31 udi=- nc=1 flg=0x8080
 n15n     (  1)      ==>newarray (in &GPR_0036) (highWordZero Unsigned profilingCode X!=0 X>=0 allocationCanBeRemoved )
------------------------------

 [     0x3fcb6346a30]                          LTG     &GPR_0324, Shadow[<contiguousArrayDataAddrField>] 8(&GPR_0036)

============================================================
...
------------------------------
 n901n    (  0)  BCDCHK [#507] (profilingCode )                                                       [     0x3fcb62e0960] bci=[2,0,790] rc=0 vc=415 vn=- li=31 udi=- nc=12 flg=0xa0
 n899n    (  1)    pdshlOverflow <prec=1 (len=1) adj=0 round=0> sign=?  (in VRF_0330) (profilingCode )  [     0x3fcb62e08c0] bci=[-1,0,8] rc=1 vc=415 vn=- li=-1 udi=- nc=2 flg=0x80
 n891n    (  0)      pdrem <prec=1 (len=1) adj=0 round=0> sign=?  (in VRF_0330) (profilingCode )      [     0x3fcb62e0640] bci=[-1,0,8] rc=0 vc=415 vn=- li=- udi=- nc=2 flg=0x80
 n886n    (  0)        pdloadi  <array-shadow>[#260  Shadow] [flags 0x80000609 0x0 ] <prec=4 (len=3) adj=0> sign=hasState  (in VRF_0326) (profilingCode )  [     0x3fcb62e04b0] bci=[-1,0,8] rc=0 vc=415 vn=- li=- udi=- nc=1 flg=0x80
 n884n    (  0)          ==>aloadi (in &GPR_0324) (profilingCode internalPtr )
 n890n    (  0)        pdloadi  <array-shadow>[#260  Shadow] [flags 0x80000609 0x0 ] <prec=2 (len=2) adj=0> sign=hasState  (in VRF_0328) (profilingCode )  [     0x3fcb62e05f0] bci=[-1,0,8] rc=0 vc=415 vn=- li=- udi=- nc=1 flg=0x80
 n888n    (  0)          ==>aloadi (in &GPR_0322) (profilingCode internalPtr )
 n28n     (  0)      ==>iconst 0 (in GPR_0179) (profilingCode X==0 X>=0 X<=0 )
 n1003n   (  0)    aloadi  <contiguousArrayDataAddrField>[#355  final Shadow +8] [flags 0x20604 0x0 ] (in &GPR_0372) (profilingCode internalPtr )  [     0x3fcb62e2940] bci=[-1,12,11] rc=0 vc=0 vn=- li=- udi=- nc=1 flg=0x8080
 n25n     (  1)      ==>newarray (in &GPR_0129) (highWordZero Unsigned profilingCode X!=0 X>=0 allocationCanBeRemoved )
 n25n     (  1)    ==>newarray (in &GPR_0129) (highWordZero Unsigned profilingCode X!=0 X>=0 allocationCanBeRemoved )
 n28n     (  0)    ==>iconst 0 (in GPR_0179) (profilingCode X==0 X>=0 X<=0 )
 n38n     (  0)    ==>iconst 1 (in GPR_0177) (profilingCode X!=0 X>=0 )
 n15n     (  0)    ==>newarray (in &GPR_0036) (highWordZero Unsigned profilingCode X!=0 X>=0 allocationCanBeRemoved )
 n28n     (  0)    ==>iconst 0 (in GPR_0179) (profilingCode X==0 X>=0 X<=0 )
 n34n     (  0)    ==>iconst 4 (in GPR_0178) (profilingCode X!=0 X>=0 )
 n20n     (  0)    ==>newarray (in &GPR_0082) (highWordZero Unsigned profilingCode X!=0 X>=0 allocationCanBeRemoved )
 n28n     (  0)    ==>iconst 0 (in GPR_0179) (profilingCode X==0 X>=0 X<=0 )
 n36n     (  0)    ==>iconst 2 (in GPR_0257) (profilingCode X!=0 X>=0 )
 n38n     (  0)    ==>iconst 1 (in GPR_0177) (profilingCode X!=0 X>=0 )
------------------------------

 [     0x3fcb63472e0]                          VLRL    VRF_0326, Shadow[<array-shadow>] 0(&GPR_0324),0x2
 [     0x3fcb6347520]                          VLRL    VRF_0328, Shadow[<array-shadow>] 0(&GPR_0322),0x1
 [     0x3fcb6347670]                          VRP     VRF_0330,VRF_0326,VRF_0328,0x1,1
...
------------------------------
 n900n    (  0)  pdstorei  <array-shadow>[#260  Shadow] [flags 0x80000609 0x0 ] <prec=1 (len=1)>  (profilingCode useStoreAsAnAccumulator )  [     0x3fcb62e0910] bci=[-1,0,8] rc=0 vc=415 vn=- li=- udi=- nc=2 flg=0x200080
 n896n    (  0)    aloadi  <contiguousArrayDataAddrField>[#355  final Shadow +8] [flags 0x20604 0x0 ] (in &GPR_0385) (profilingCode internalPtr )  [     0x3fcb62e07d0] bci=[2,131,805] rc=0 vc=432 vn=- li=31 udi=- nc=1 flg=0x8080
 n25n     (  0)      ==>newarray (in &GPR_0129) (highWordZero Unsigned profilingCode X!=0 X>=0 allocationCanBeRemoved )
 n899n    (  0)    ==>pdshlOverflow <prec=1 (len=1) adj=0 round=0> sign=?  (in VRF_0330) (profilingCode )
------------------------------

 [     0x3fcb634d9f0]                          LTG     &GPR_0385, Shadow[<contiguousArrayDataAddrField>] 8(&GPR_0129)
 [     0x3fcb634dac0]                          VSTRL   VRF_0330, Shadow[<array-shadow>] 0(&GPR_0385),0x0
```

Comment (VermaSh) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2108]: Putting my above comment aside; BCDCHK IL will never have an address node warped inside a `aloadi`. So shouldn't we be good evaluating offset trees like we do in this PR?

Comment (zl-wang) [reply runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp:2108]: for this particular spot, yes, i think it is ok, since MemoryReference is not used on that `aladd`.

**Review by zl-wang — APPROVED** [review]
with the caveat: in case aRegLoad happening, it will be fixed then (be it due to registers invisible to dependency build-up or code skipped over narrow checking -- in my mind, indeed that existed in other codegens too).

**Comment by r30shah** [comment]
jenkins test sanity zlinux jdk17,jdk21

**Comment by VermaSh** [comment]
Fixed variable name causing the build failure. @r30shah Can you please relaunch the tests

**Comment by r30shah** [comment]
jenkins test sanity zlinux jdk17,jdk21

**Comment by r30shah** [comment]
Jenkins test sanity.system zlinux jdk21

**Comment by r30shah** [comment]
Merging this change as all the tests passed.

**Comment by r30shah** [comment]
@VermaSh  We would need to backport this to 0.51, can you open up the PR against it so that once this propagates , it can be merged ?
